app/navigation/navigator.py

```diff
@@ -1,82 +1,156 @@
-from app.utilities.factory import factory
 import logging
-from app.model.block import Block
+from app.navigation.navigation_store import NavigationStore
+from app.submitter.converter import SubmitterConstants
+from app.utilities.factory import factory
+from flask import session
+from flask_login import current_user
+from app.submitter.submitter import SubmitterFactory
+from app import settings
 
 
 logger = logging.getLogger(__name__)
 
 
-class NavigationException(Exception):
+class UserActionException(Exception):
     pass
 
 
-class Navigator(object):
-    def __init__(self, schema, navigation_history):
+class UserAction(object):
+    '''
+    Using the chain of responsibility pattern hand off a user action down the chain
+    until one of the objects can handle it
+    '''
+    def __init__(self, schema, metadata):
         self._schema = schema
-        self._navigation_history = navigation_history
-        self._store = factory.create("navigation-store")
+        self._metadata = metadata
+        self.next_action = None
+
+    def set_next_action(self, action):
+        self.next_action = action
+        return action
+
+    def process_action(self, action, current_location):
+        # work through the user actions to determine which one matches the request action
+        if action == self.get_action():
+            # perform any functionality necessary to complete that action
+            self.perform_action()
+            # then provide the next location
+            return self.provide_next_location(current_location)
+        else:
+            # request the next location
+            return self.next_action.process_action(action, current_location)
 
-    # destination  "group:block:section:question:<repetition>"
-    def _valid_destination(self, destination):
-        """
-        Returns a boolean indicating whether a proposed destination is a valid one
+    def get_action(self):
```

Comment:
Does this getter always return a string? If so, is it not a property that is set rather than a method?